src/v/utils/vint.h

```diff
@@ -12,8 +12,109 @@
 #pragma once
 #include "bytes/bytes.h"
 
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
 #include <cstdint>
 
+namespace unsigned_vint {
+/// At most 5 bytes are needed to encode a 32 bit value
+inline constexpr size_t max_length = 5;
+
+namespace detail {
+struct var_decoder {
+    uint64_t result = 0;
+    size_t bytes_read = 0;
+    uint64_t shift = 0;
+    uint8_t limit;
+
+    explicit var_decoder(uint8_t limit)
+      : limit(limit) {}
+
+    /**
+     * @brief Incorporate a new byte into the state.
+     * Returns true if the outer loop should break (last byte found).
+     */
+    bool accept(uint8_t byte) {
+        /// \brief almost identical impl to leveldb, made generic for c++
+        /// friendliness
+        /// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
+        bytes_read++;
+        if (byte & 128) {
+            result |= ((byte & 127) << shift);
+        } else {
+            result |= byte << shift;
+            return true;
+        }
+        shift += 7;
+        return shift > limit;
+    }
+};
+
+template<typename Range>
+inline std::pair<uint64_t, size_t>
+deserialize(Range&& r, uint8_t limit) noexcept {
+    var_decoder decoder(limit);
+    for (auto byte : r) {
+        if (decoder.accept(static_cast<uint8_t>(byte))) {
+            break;
+        }
+    }
+    return std::make_pair(decoder.result, decoder.bytes_read);
+}
+
+} // namespace detail
+
+inline ss::future<std::pair<uint32_t, size_t>>
+deserialize(ss::input_stream<char>& s) {
+    detail::var_decoder decoder(5 * sizeof(uint64_t));
+    while (!s.eof()) {
+        auto buf = co_await s.read_exactly(1);
+        if (decoder.accept(static_cast<uint8_t>(buf[0]))) {
+            break;
+        }
+    }
+    co_return std::make_pair(
+      static_cast<uint32_t>(decoder.result), decoder.bytes_read);
+}
+
+inline size_t serialize(uint64_t value, uint8_t* out) noexcept {
+    size_t bytes_used = 0;
```

Comment:
i could have sworn we already have this code i remember typing something like this and benchmarking the impl and branches.

Reply:
ah, you moved parts of it.